### DISCUSSION TITLE
replacing the __call__ function and creating a new type instance on Pyo3 side in Rust

### DISCUSSION THREAD

**Discussion by markusdd** [PyO3/pyo3#3380] [Questions]
Ok, so this is some major metaproogramming question. Rationale:

In our project we have a plethora of type/class instances which have the same signature and are frozen after creation, hence the instances can be shared. But an upfront  singleton type approach is not really practical for our purposes. For that there exists already a Python package called mementos which implements the Memento-pattern:
https://pypi.org/project/mementos/

It is relatively simple: It creates a dict key out of the class and it's calling signature, saves the instance in a dict, and overrides the call function of a class that inherits the metaclass menentos. If the class is being created somewhere else with the same signature, the same instance is returned again instead of creating a new one.

As this inital creation and lookup is very performance critical in our project we want to try if we can gain anything by essentially re-implenting this in rust. I just tried to replace the key generation functioon and this actually made it slower, soo now I am looking into moving the whole dict into a rust hashmap and just handing in the signature and retuning a new or cached instance, so essentially I want to oxidize this call function from the mementos package, alternatively one could also replace the whole outer metaclass factory but that might not be worth it:

```py
_memento_cache = {}

def memento_factory(name, func):
    """
    Return a memoizing metaclass with the given name and key function.
    That makes this a parametrized meta-metaclass, which is probably
    the most meta thing you've ever seen. If it isn't, both congratulations
    and sympathies are in order!
    """

    def call(cls, *args, **kwargs):
        key = func(cls, args, kwargs)
        try:
            return _memento_cache[key]
        except KeyError:
            instance = type.__call__(cls, *args, **kwargs)
            _memento_cache[key] = instance
            return instance

    mc = type(name, (type,), {"__call__": call})
    return mc

```

My problem are the calls to 'type'. I cannot really find the equivalent for that in pyoo3, oor I misunderstand the docs.
Or can I just create the instance based on cls which is handed to me?
If I write cls.__call__(*args, **kwargs) in python that will lead to infinite recursion.

Any tips are appreciated.

**Comment by davidhewitt** [comment]
Could you fetch `type` from builtins? e.g. `py.import("builtins")?.getattr("type)?`

**Reply by markusdd** [reply]
I noticed in the FFI docs that there is PyTypeObject and PyTypeInfo but basically without any info. Would your way allow me to call __call__ on it?

**Reply by davidhewitt** [reply]
Yes, e.g. `.call1((name, (type,), d))` where `d` is a dict with `{"__call__": call}`.

**Reply by markusdd** [reply]
Ah call1 is the secret. I don't know why I misinterpreted what it says it does.

Will try, thanks a ton!

**Reply by markusdd** [reply]
mmh...facing another issue.

I need too keep the hashmap as part oof the mdooule and somehow hand it into the call function, but hoow to do that withoout exposing this argument as part of the interface into python?

**Reply by davidhewitt** [reply]
Probably at the moment you are looking for `GILOnceCell` to keep the hashmap in a `static` variable. In the future I hope to support private module state.

**Reply by markusdd** [reply]
ok will try to look into that...sorry for all the 'o's my keyboard is acting up...

**Reply by markusdd** [reply]
mmh...that kind of makes sense to me for python types but is there no way to maintain a completely rust internal HashMap?

Also, is this HashMap then still mutable? I mean it has to change if a new instance must be added.

**Reply by davidhewitt** [reply]
Maybe `GILProtected` or some other mutex type is a better fit then.

**Reply by markusdd** [reply]
Ok so I made some good progress overall on this, but now I am struggling to implement the shared Hashmap or Dict:

![grafik](https://github.com/PyO3/pyo3/assets/25175069/4e999334-cd73-44e6-8046-0074c88481ba)

Creating the key and value for the Hashmap fails like this...any ideas how to handle this on the rust side? I mean this is what the types are...somehow it must work in Python, but how to do this in Rust?

**Reply by markusdd** [reply]
So using the OnceCell and a PyDict at least seems to work, now I only have issues with the instance creation Oo
![grafik](https://github.com/PyO3/pyo3/assets/25175069/ec3d324c-6574-4d48-b88c-7e16e413fad2)

**Reply by davidhewitt** [reply]
Try `pytype.call(comb_args, kwargs)` instead of trying to go through `__call__` manually

**Reply by adamreichold** [reply]
> Ok so I made some good progress overall on this, but now I am struggling to implement the shared Hashmap or Dict:

If you want to store these globally, you need to use the detached pointers into the Python heap, e.g.

```rust
HashMap<(Py<PyAny>, Py<PyAny>, Py<PyAny>), Py<PyPany>>
```

as references like `&'py PyAny` are bound to holding the GIL for at least `'py`.

**Reply by markusdd** [reply]
Thanks for that, good point. Will revisit the HashMap later for benchmarking, but currently I am stuck here:
![grafik](https://github.com/PyO3/pyo3/assets/25175069/0e9fa46f-a6d4-4095-807b-deeb7845cf46)

the tip from @davidhewitt was correct to just use call, but now it complains that 'tape' has 1 or 3 arguments (python stacktrace). So I seperated cls and args again, but now rust complains call only takes 2 xD

By the way thanks guys for helping here, maybe I should've have taken such metaprogramming stuff as my intro to pyo3...but hey...you always learn I guess.

**Reply by davidhewitt** [reply]
Sounds like really you wanted `pytype.call1((cls, args, kwargs))`?

**Reply by markusdd** [reply]
Ok that is wild^^
This indeed works, but I never would have guessed from the docs that this would be the correct approach. I thought call1 cannot contain kwargs, but as this is 'meta' the kwargs are just another arg I guess 😅 
![grafik](https://github.com/PyO3/pyo3/assets/25175069/b0d17491-e6f5-407b-9b68-3178df087029)


Now we are in some interesting territory though because it still does not seem to be equivalent to the original Python code:
![grafik](https://github.com/PyO3/pyo3/assets/25175069/0ae6ba01-60bf-401a-b456-7c77139c93ed)

**Reply by markusdd** [reply]
Also, it turns out `type.__call__(...)` is not equivalent to `type(...)`,so I think somehow we need to get call_method to work...
![grafik](https://github.com/PyO3/pyo3/assets/25175069/993fb158-aee5-4cb1-bcd3-ec6c117cb2e1)

**Reply by markusdd** [reply]
I have verified now that Python passes in the cls argument as a string:
![grafik](https://github.com/PyO3/pyo3/assets/25175069/390f5103-fac9-441a-90ec-d7f07a4a57c3)

If that is the case, how does it work on the python side? Is there a cast that can be performed to turn this string into 'a type object' based on the name that is represented by the string?

**Reply by markusdd** [reply]
What is happening here?
![grafik](https://github.com/PyO3/pyo3/assets/25175069/936ae40c-8a8f-40c0-bbf6-b90af4373160)

When I check the type of cls in python it is the metaclass instead of some string?!
Why would that change once this is passed into Rust?

I need some help here, that does not make sense to me.

**Reply by markusdd** [reply]
Ok, now we are entering clown world.
If I do it like this it works with the rust implementation, but is super slow.
This makes 0 sense to me.
![grafik](https://github.com/PyO3/pyo3/assets/25175069/0a168d21-ac51-4040-b769-f40ea16219a0)

Here is the version where I give the rust function directly. This must be some kind of bug or some weirdness with the signtures.
![grafik](https://github.com/PyO3/pyo3/assets/25175069/f1a1ded2-13d2-44a8-99af-6d8d6b6d97bd)

Also: why is this so horribly slow?

**Reply by markusdd** [reply]
any ideas @davidhewitt @adamreichold what could explain this? I experimented a bit more and I do not understand why the argument type changes when I nest the functions and also why the rust version is so slow. That does not really make sense to me right now.

**Reply by davidhewitt** [reply]
Are you building with `--release`? Also, my intuition is that rewriting this function in Rust isn't going to yield much performance speedup, because it's a relatively light Python function and the Rust implementation is heavily calling into Python anyway.

Regarding the difference between Python and Rust `cls` type, I am afraid I don't have an answer there. The details may lie in the implementation of `type()`.

**Reply by markusdd** [reply]
yes I am building with --release.

You might be correct that no substantial speedup is to be expected here, but seeing a 6x slowdown is equally unexpected.

My issue with the cls datatype is that in my view just nesting it into another function does actually absolutely nothing, that is what I find utterly confusing.

One other idea: can the import of type be pulled out to be permanent?
I could imagine reimporting on every call might be expensive.

**Reply by birkenfeld** [reply]
The answer here lies in the descriptor behavior (`__get__`) that Python defined functions have, and that "C functions", which PyO3 also uses, lack. The latter cannot be used as methods without being specially wrapped by being defined in the `PyMethodDef` list for the class.

The C functions do not get implicitly passed the instance (normally `self`, here `cls`) as the first argument. Therefore, in the Rust function what you're getting as `cls` is simply the first argument being split from `*args`: the class name as a string.

**Reply by birkenfeld** [reply]
Here is a simple PyO3-free example demonstrating this:

```python
import _testcapi

def py_meth(*args):
    return args

c_meth = _testcapi.get_args

print(py_meth(1, 2, 3))
print(c_meth(1, 2, 3))

py_type = type("A", (object,), {"__call__": py_meth})
c_type = type("B", (object,), {"__call__": c_meth})

print("---")
print(py_type()(1, 2, 3))
print(c_type()(1, 2, 3))
```

(`_testcapi.get_args` is a simple test C function defined just like `py_meth` is here: it returns all its positional arguments as a tuple.)

Output:
```
(1, 2, 3)
(1, 2, 3)
---
(<__main__.A object at 0x7faa654e6e90>, 1, 2, 3)
(1, 2, 3)
```

**Reply by markusdd** [reply]
@birkenfeld You sir just blew my mind haha

But it makes complete sense of course. I once learned this when I started using Python more than a decade ago but as this is somethign you normally never bother with it kind of did not come to my mind when debugging this.

Question is: would there be a way to circumvent this? Or is it anyway not worth the effort?

You mentioned PyMethodDef...can we inject the rust call function in a more effective manner than just wrapping it into a python function?

Or is there a way to completely replace the whole python mementos module?
I am not quite sure how I would handle creation of the metaclass, but if this would work we would surely save some calling overhead.
